### PR TITLE
bootctl: Add set-entry verb to create a boot loader entry in $BOOT/loader/entries/

### PR DESCRIPTION
**Pull request by VasilisMylonas**
Add a set-entry verb to the bootctl command to allow creating a boot loader entry in $BOOT/loader/entries/ . This implementation discovers as much info as possible by using /etc/os-release and other relevant files. The currently running kernel and initrd can also be discovered sometimes, otherwise they may be provided as arguments.

### CODE REVIEW

**Comment by poettering** [comment]
humm, what's the usecase here? The assumption was that the "kernel-install" with its plugins is responsible for this. Why replicate that in bootctl?

**Comment by VasilisMylonas** [comment]
I thought the purpose of kernel-install is to install the actual kernel/initrd to a location (/boot for example). From what I see it can indeed create boot entries but shouldn't boot entry management be controlled by bootctl? I believe this could also be extended to arbitrary EFI executables not just the Linux kernel/initrd.

File: src/boot/bootctl.c [hunk 1]
```diff
@@ -2555,6 +2556,334 @@ static int verb_reboot_to_firmware(int argc, char *argv[], void *userdata) {
...
+        if (entry->options) {
+                for (size_t i = 0; i < SIZE_MAX; i++)
+                {
+                        if (!entry->options[i])
+                                break;
+
+                        free(entry->options[i]);
+                }
+        }
+        free(entry->options);
```
Comment (keszybz) [inline src/boot/bootctl.c]: That looks like a reimplementation of strv_free().

File: src/boot/bootctl.c [hunk 2]
```diff
@@ -2555,6 +2556,334 @@ static int verb_reboot_to_firmware(int argc, char *argv[], void *userdata) {
...
+        char* s = strjoin("linux    ", entry->kernel, "\n");
+
+        if (entry->title) {
+                s = strpush(s, "title    ", entry->title, "\n");
+                if (!s) {
+                        free(s);
+                        return NULL;
+                }
+        }
+
```
Comment (keszybz) [inline src/boot/bootctl.c]: Each append here copies the string, so it scales as the square of the number of strings added. We generally don't have too many strings, but this is still pretty ugly. `open_memstream_unlocked()` or a normal file should be used instead.

File: src/boot/bootctl.c [hunk 3]
```diff
@@ -2555,6 +2556,334 @@ static int verb_reboot_to_firmware(int argc, char *argv[], void *userdata) {
...
+        if (!os_release_version) {
+                os_release_version = strdup("No Version Information");
+                if (!os_release_version) {
+                        r =-ENOMEM;
+                        goto error;
+                }
+        }
```
Comment (keszybz) [inline src/boot/bootctl.c]: It doesn't make sense to fill in fallback values like this. If something is missing, then it should be missing in the menu entry too. Every user of menu entries is supposed to implement suitable fallbacks.

File: src/boot/bootctl.c [hunk 4]
```diff
@@ -2555,6 +2556,334 @@ static int verb_reboot_to_firmware(int argc, char *argv[], void *userdata) {
...
+
+        char* file_name;
+
+        if (kernel_release != NULL)
+                file_name = strjoina(machine_id, "-", kernel_release, ".conf");
+        else
+                file_name = strjoina(machine_id, ".conf");
+
+        return path_join(arg_dollar_boot_path(), "loader/entries", file_name);
+}
```
Comment (keszybz) [inline src/boot/bootctl.c:2691]: This is missing the support for `entry-token` that was added within the last few months. Machine-id is only used as a fallback now.

File: src/boot/bootctl.c [hunk 5]
```diff
@@ -2571,7 +2900,7 @@ static int bootctl_main(int argc, char *argv[]) {
                 { "random-seed",         VERB_ANY, 1,        0,            verb_random_seed         },
                 { "systemd-efi-options", VERB_ANY, 2,        0,            verb_systemd_efi_options },
                 { "reboot-to-firmware",  VERB_ANY, 2,        0,            verb_reboot_to_firmware  },
-                {}
```
Comment (keszybz) [inline src/boot/bootctl.c:2574]: This line terminates the table and cannot be removed.

**Review by keszybz — COMMENTED** [review]
Hmm, the code has many issues, but I don't think it makes sense to dive into them before we figure out what the patch is trying to do.

Please describe what the new verb does, and when it would be used.
Also explain why `kernel-install` is not appropriate.

**Comment by VasilisMylonas** [comment]
@keszybz  The new verb serves a similar purpose to kernel-install but should be more flexible in order to allow arbitrary configurations and not just the kernel/initrd. It could for example be used to add an entry for a different boot manager or any other EFI (for example the EFI version of gptfdisk). Because boot entry management is not really specific to the kernel I thought it would be more fitting for bootctl to be able to create entries, seeing as it already has verbs like list and set-default. I thought it would be nice to also have a add-entry, remove-entry or something similar. I appreciate the constructive criticism and definitely agree that this needs more work (if this is to be added anyways).

File: src/boot/bootctl.c (hunk 1 above)
Comment (VasilisMylonas) [reply src/boot/bootctl.c]: Correct I missed that, new commit fixes it.

File: src/boot/bootctl.c (hunk 2 above)
Comment (VasilisMylonas) [reply src/boot/bootctl.c]: Just changed it to use a caller-provided stream.

**Pull request closed without merging**